### PR TITLE
Allow setting version explicitly

### PR DESCRIPTION
**Pull request by PlugFox**
@f3ath 
#9 

Also added many cosmetic improvements

### CODE REVIEW

File: .github/workflows/dart.yml [hunk 1]
```diff
@@ -2,9 +2,9 @@ name: Dart CI
 
 on:
   push:
-    branches: [ master ]
+    branches: [ master, dev ]
```
Comment (f3ath) [inline .github/workflows/dart.yml:5]: I don't think there is a `dev` branch.

File: .gitignore [hunk 2]
```diff
@@ -19,3 +19,6 @@ doc/api/
 *.js_
 *.js.deps
 *.js.map
+
+# IDEA
+.idea/
```
Comment (f3ath) [inline .gitignore:24]: This project has nothing to do with IDEA in particular. Developer-specific stuff belongs in the personal global `.gitignore`.

File: lib/cider.dart [hunk 3]
```diff
@@ -1,3 +1,11 @@
-export 'package:cider/src/console/console.dart';
-export 'package:cider/src/console/console_application.dart';
-export 'package:cider/src/console/exit_code.dart';
+library cider;
+
+export 'package:cider/src/console/console.dart'
+    // ignore: uri_does_not_exist
+    if (dart.library.html) 'package:cider/src/web_support_stub.dart';
```
Comment (f3ath) [inline lib/cider.dart:5]: Why is this needed? Do we intend to run in a browser?

File: lib/src/application.dart [hunk 4]
```diff
@@ -48,6 +48,11 @@ class Application {
     _writeChangelog(changelog);
   }
 
+  String set(Version version) {
+    _writeVersion(version);
+    return version.toString();
```
Comment (f3ath) [inline lib/src/application.dart:53]: What is the point in returning the argument. The caller logic has it already, right?

File: lib/src/console/command/set_command.dart [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      } else {
+        Version version;
+        try {
+          version = Version.parse(argResults.rest.single);
+        } on dynamic catch (_) {
+          throw UnsupportedError('Can\'t parse version');
+        }
+        _console.log(createApp().set(version));
+      }
+    } on UnsupportedError catch (e) {
```
Comment (f3ath) [inline lib/src/console/command/set_command.dart:36]: Error management logic is currently [here](https://github.com/f3ath/cider/blob/master/bin/cider.dart#L12). I'd prefer to keep it consolidated.

File: lib/src/console/command/set_command.dart [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+      _console.error(e?.message ?? 'An unsupported error occurred');
+      return ExitCode.usageException;
+    } on dynamic catch (e) {
```
Comment (f3ath) [inline lib/src/console/command/set_command.dart:39]: I'm not sure I understand the expected behavior here. Is there a supporting test?

File: lib/src/console/command/set_command.dart [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    try {
+      if (argResults.rest.isEmpty) {
+        throw UnsupportedError('Version must be specified');
+      } else if (argResults.rest.length > 1) {
+        throw UnsupportedError('Too many arguments passed');
+      } else {
+        Version version;
+        try {
+          version = Version.parse(argResults.rest.single);
+        } on dynamic catch (_) {
```
Comment (f3ath) [inline lib/src/console/command/set_command.dart:31]: I think `Version.parse` throws `FormatException` in this case. What's the reason for using `dynamic`?

File: lib/src/console/console.dart [hunk 8]
```diff
@@ -4,16 +4,20 @@ import 'dart:io';
 /// Allows to print log messages and error messages.
 class Console {
   /// Creates an instance with
-  const Console(this._output, this._error);
+  const Console(this._output, this._error, this._supportsAnsiEscapes);
```
Comment (f3ath) [inline lib/src/console/console.dart:7]: `Console` is exported by Cider, hence it is a part of external API (i.e. contract). Let's make the third argument optional, otherwise this technically makes a BC-breaking change.

Comment (f3ath) [reply lib/src/console/console.dart:7]: Actually disregard the first comment. Why do we need this argument at all, it should be possible to read it from `this.stdout`, shouldn't it?

File: pubspec.yaml [hunk 9]
```diff
@@ -1,7 +1,9 @@
 name: cider
-version: 0.0.4
+version: 0.0.4-allow-setting-version-explicitly
```
Comment (f3ath) [inline pubspec.yaml:2]: `0.0.5` should be OK

File: pubspec.yaml [hunk 10]
```diff
@@ -18,7 +20,8 @@ dependencies:
 
 dev_dependencies:
   pedantic: ^1.9.0
-  test: ^1.9.0
+  test: ^1.15.3
```
Comment (f3ath) [inline pubspec.yaml:23]: Why did this get downgraded?

File: .github/workflows/dart.yml [hunk 11]
```diff
@@ -20,5 +20,17 @@ jobs:
         run: pub get
       - name: Analyzer
         run: dartanalyzer lib test --fatal-infos --fatal-warnings
-      - name: Tests
-        run: pub run test
+      - name: Format
+        run: dartfmt --dry-run --set-exit-if-changed lib
+      - name: Run tests
+        run: pub run test_coverage --no-badge --print-test-output --min-coverage 0
+      #- name: Upload coverage to Codecov
```
Comment (f3ath) [inline .github/workflows/dart.yml:27]: Do we need the commented-out part?

File: .github/workflows/dart.yml [hunk 12]
```diff
@@ -20,5 +20,17 @@ jobs:
         run: pub get
       - name: Analyzer
         run: dartanalyzer lib test --fatal-infos --fatal-warnings
-      - name: Tests
-        run: pub run test
+      - name: Format
+        run: dartfmt --dry-run --set-exit-if-changed lib
+      - name: Run tests
+        run: pub run test_coverage --no-badge --print-test-output --min-coverage 0
```
Comment (f3ath) [inline .github/workflows/dart.yml:26]: Not sure I understand the reason for `--min-coverage 0`

File: README.md [hunk 13]
```diff
@@ -1,5 +1,10 @@
-![logo]
+![logo]
```
Comment (f3ath) [inline README.md:1]: extra spaces?

File: README.md [hunk 14]
```diff
@@ -38,6 +43,17 @@ cider log add 'Support for rocket fuel and kerosene'
 cider log fix 'No more wheels falling off'
 ```
 
+## Setting the project version
+```
+cider set <version>
```
Comment (f3ath) [inline README.md:48]: My initial vision was to combine this with the existing `version` command. E.g.
- `cider version` prints the version
- `cider version 1.2.3` sets the version

What do you think?

File: lib/cider.dart (hunk 3 above)
Comment (PlugFox) [reply lib/cider.dart:5]: This is for pub pana rating)

File: lib/src/application.dart (hunk 4 above)
Comment (PlugFox) [reply lib/src/application.dart:53]: Return signature is the same as `bump` method.

File: .github/workflows/dart.yml (hunk 1 above)
Comment (PlugFox) [reply .github/workflows/dart.yml:5]: But it would be useful.
For example, I wouldn't have to do PR in master)

File: .github/workflows/dart.yml (hunk 12 above)
Comment (PlugFox) [reply .github/workflows/dart.yml:26]: Required to upload results to Codecov.

Check the pipeline at the end with `ChicagoFlutter/lcov-cop@v1.0.2`

```
      #- name: Upload coverage to Codecov
      #  uses: codecov/codecov-action@v1
      #  with:
      #    token: ${{ secrets.CODECOV_TOKEN }}
      #- name: Check Code Coverage
      #  uses: ChicagoFlutter/lcov-cop@v1.0.2
      #  with:
      #    path: coverage/lcov.info
      #    min_coverage: 100
      #    exclude: "**/*.g.dart"
```

File: .github/workflows/dart.yml (hunk 11 above)
Comment (PlugFox) [reply .github/workflows/dart.yml:27]: Yeap, we need.

It's free service for badge and graphical representation of coverage.
https://codecov.io

File: .gitignore (hunk 2 above)
Comment (PlugFox) [reply .gitignore:24]: I add things like this to local `.gitignore`)

File: README.md (hunk 13 above)
Comment (PlugFox) [reply README.md:1]: Psychological trauma from working with Github's markdown.

File: lib/cider.dart (hunk 3 above)
Comment (PlugFox) [reply lib/cider.dart:5]: To get 110 out of 110.

File: lib/src/console/command/set_command.dart (hunk 7 above)
Comment (PlugFox) [reply lib/src/console/command/set_command.dart:31]: Good point, I didn't know about the type of error, so I did dynamic.

File: lib/src/console/command/set_command.dart (hunk 5 above)
Comment (PlugFox) [reply lib/src/console/command/set_command.dart:36]: I have styled the error output to the console C:

Also `runZoned` in main much handier and prettier for catching global errors, in my opinion)

File: lib/src/console/command/set_command.dart (hunk 6 above)
Comment (PlugFox) [reply lib/src/console/command/set_command.dart:39]: Not all UnsupportedError have 'message'. Nullable field.

```
/**
 * The operation was not allowed by the object.
 *
 * This [Error] is thrown when an instance cannot implement one of the methods
 * in its signature.
 */
@pragma("vm:entry-point")
class UnsupportedError extends Error {
  final String? message; // <= Nullable field.
  @pragma("vm:entry-point")
  UnsupportedError(String this.message);
  String toString() => "Unsupported operation: $message";
}
```

File: lib/src/console/console.dart (hunk 8 above)
Comment (PlugFox) [reply lib/src/console/console.dart:7]: I didn't think it was part of the API.
This field is needed for formatted output to the console.

File: pubspec.yaml (hunk 9 above)
Comment (PlugFox) [reply pubspec.yaml:2]: My idea is that you fix things yourself, and then bump version c:

File: pubspec.yaml (hunk 10 above)
Comment (PlugFox) [reply pubspec.yaml:23]: Upgraded, not downgraded)

File: README.md (hunk 14 above)
Comment (PlugFox) [reply README.md:48]: It's ok)
You seem to have support for command aliases.

File: .github/workflows/dart.yml (hunk 1 above)
Comment (f3ath) [reply .github/workflows/dart.yml:5]: Changing branch structure is out of scope of this PR. Let's keep the concerns separated.

File: .gitignore (hunk 2 above)
Comment (f3ath) [reply .gitignore:24]: So what is the reason for adding it here?

File: lib/cider.dart (hunk 3 above)
Comment (f3ath) [reply lib/cider.dart:5]: I prefer clean an concise code over the pub rating, tbh. Let's optimize towards 100% test coverage instead.

File: lib/src/application.dart (hunk 4 above)
Comment (f3ath) [reply lib/src/application.dart:53]: Right, but bump returns something which is not known beforehand. In this case the returned value is always known ahead.

File: lib/src/console/command/set_command.dart (hunk 6 above)
Comment (f3ath) [reply lib/src/console/command/set_command.dart:39]: Still unclear what it has to do with `} on dynamic catch (e) {`. Can you elaborate please?

File: lib/src/console/console.dart [hunk 15]
```diff
@@ -4,16 +4,20 @@ import 'dart:io';
...
+
+  final bool _supportsAnsiEscapes;
 
   final Stdout _output;
 
   final Stdout _error;
 
   /// Writes the [message] to the error sink
-  void error(Object message) => _error.writeln(message.toString());
+  void error(Object message) => _supportsAnsiEscapes
```
Comment (f3ath) [inline lib/src/console/console.dart:18]: Can we change this line to `void error(Object message) => _output.supportsAnsiEscapes` ?

File: lib/src/console/console.dart (hunk 8 above)
Comment (f3ath) [reply lib/src/console/console.dart:7]: See the comment below.

File: pubspec.yaml (hunk 10 above)
Comment (f3ath) [reply pubspec.yaml:23]: Oh by bad.

**Comment by f3ath** [comment]
Closing as partially merged
